nighthawk/source/common/utility.cc

```diff
@@ -29,26 +29,73 @@ uint32_t determineCpuCoresWithAffinity() {
 
 } // namespace PlatformUtils
 
-Uri::Uri(const std::string& uri) : scheme_("http") {
+size_t Uri::findPortSeparatorInAuthority(absl::string_view authority) {
+  size_t colon_index = std::string::npos;
+  bool in_ipv6_address = false;
+  bool last_seen_dot = true;
+
+  // TODO(oschaaf): We need to revisit this and improve. This performs some
+  // limited hostname validation.
+  for (size_t i = 0; i < authority.size(); i++) {
+    char c = authority[i];
+    if (i == 0 && c == '[') {
+      in_ipv6_address = true;
+    } else if (in_ipv6_address && c == ']') {
+      in_ipv6_address = false;
+    } else {
+      if (!in_ipv6_address && i > 0 && c == ':') {
+        colon_index = i;
+        break;
+      } else if (c == '.') {
+        if (last_seen_dot) {
+          throw InvalidHostException("Invalid hostname");
+        }
+        last_seen_dot = true;
+        continue;
+      } else {
+        last_seen_dot = false;
+      }
+
+      bool ok = std::isalnum(c) || (i > 0 && c == '-') || (in_ipv6_address && c == ':');
+
+      if (!ok) {
+        throw InvalidHostException("Invalid hostname");
+      }
+    }
+  }
+  if (in_ipv6_address) {
+    throw InvalidHostException("Invalid hostname (missing']')");
+  }
+
+  return colon_index;
+}
+
+Uri::Uri(absl::string_view uri) : scheme_("http") {
   absl::string_view host, path;
   Envoy::Http::Utility::extractHostPathFromUri(uri, host, path);
   host_and_port_ = std::string(host);
   path_ = std::string(path);
-  const size_t colon_index = host_and_port_.find(':');
   const bool is_https = absl::StartsWith(uri, "https://");
   const size_t scheme_end = uri.find("://", 0);
-
   if (scheme_end != std::string::npos) {
     scheme_ = absl::AsciiStrToLower(uri.substr(0, scheme_end));
   }
 
+  size_t colon_index;
+
+  try {
+    colon_index = findPortSeparatorInAuthority(host_and_port_);
+  } catch (InvalidHostException ex) {
+    host_error_ = true;
```

Comment:
Why not throw an exception in the constructor? This is the idiomatic C++ way to fail a constructor.